### ISSUE TITLE
Use 3rd party HTTP parsing library?

### ISSUE THREAD

**Issue by DimitriPapadopoulos** [adrienverge/openfortivpn#347]
Would it make sense to use a 3rd party HTTP parsing library instead of maintaining our own code?

Examples of 3rd party libraries:
* [http-parser](https://github.com/nodejs/http-parser) - no syscalls or memory allocation
* [picohttpparser](https://github.com/h2o/picohttpparser) - no memory allocation, ×5 faster than [http-parser](https://github.com/nodejs/http-parser)

**Comment by mrbaseman** [comment]
hmm... for what reason? 
* more features / flexibility that a 3rd party library provides? 
* issues we have/had with our own code which are properly handled there (if so, which ones?)? 
* just to shrink our code and do outsourcing of this particular task? 

personally, if the code for a particular feature is not too large, I prefer to maintain it locally for the following reasons:

* it's probably easier to understand and debug than a "black box" that is supposed to do a particular task
* we stay flexible to extend it on demand according to our requirements
* it's more portable than external libraries that might not exist on some platform

I'm not strictly against 3rd party libraries, just expressing my vague feelings about this topic

**Comment by DimitriPapadopoulos** [comment]
Only this one:
* just to shrink our code and do outsourcing of this particular task

Both HTTP parsers are at least as portable as _openfortivpn_. Just and idea though, not a necessity.

**Comment by mrbaseman** [comment]
hmm... we would replace (large) parts of [http.c](https://github.com/adrienverge/openfortivpn/blob/master/src/http.c). But we would presumably bundle the parser with _openfortivpn_, and redistribute it together with our source, right?

**Comment by DimitriPapadopoulos** [comment]
Yes, that's what I had in mind. Linux packages might want to link to a library though.

**Comment by mrbaseman** [comment]
good point: if the external library provides more detailed error messages. On the other hand one could just print out `res` to stderr before freeing it [here](https://github.com/adrienverge/openfortivpn/blob/2aca389/src/http.c#L571), at least in the verbose mode. 
But maybe this is just one more advantage of an external library, and there are many more if we just look close enough...

**Comment by mrbaseman** [comment]
string parsing could also be done by an external library (see #502)

**Comment by DimitriPapadopoulos** [comment]
#575: one more reason to use external files/libraries for everything that is not core functionality: HTTP or XML parsing

**Comment by DimitriPapadopoulos** [comment]
#642 cleans up _http_receive()_, preparing the way for a possible [PicoHTTPParser](https://github.com/h2o/picohttpparser) integration. At this point it's not yet clear if a foreign HTTP parsing library will result in less and more efficient code. The HTTP parsing code is currently specific  and goal-oriented, hence quite efficient and simple. Not certain more general 3rd party code will be better.
